### ISSUE TITLE
PUT of $ref on navigation property is throwing exception, GET on /accounts?$expand=primarycontactid/$ref doesn't work

### ISSUE THREAD

**Issue by kalavany** [OData/odata.net#762]
/accounts(c215ae71-9ff8-e611-80d5-00155d224f03)/primarycontactid/$ref this throws the below exception - 

{
  "error": {
    "code": "",
    "message": "No HTTP resource was found that matches the request URI 'http://kalavany3/test/api/data/v8.2/accounts(c215ae71-9ff8-e611-80d5-00155d224f03)/primarycontactid/$ref'."
  }
}

GET on /accounts?$expand=primarycontactid/$ref
Fails with exception -
{
"error": {
"code": "0x0",
"message": "Multiple properties with the name 'primarycontactid' were detected in a resource or a complex value. In OData, duplicate property names are not allowed.",
"innererror": {
"message": "Multiple properties with the name 'primarycontactid' were detected in a resource or a complex value. In OData, duplicate property names are not allowed.",
"type": "Microsoft.OData.ODataException",
"stacktrace": " at Microsoft.OData.DuplicatePropertyNameChecker.ValidatePropertyUniqueness(ODataNestedResourceInfo property)\r\n at Microsoft.OData.ODataWriterCore.<>c__DisplayClass91_0.b__1()\r\n at Microsoft.OData.ODataWriterCore.InterceptException(Action action)\r\n at Microsoft.OData.ODataWriterCore.CheckForNestedResourceInfoWithContent(ODataPayloadKind contentPayloadKind, ODataItem contentPayload)\r\n at Microsoft.OData.ODataWriterCore.WriteStartResourceImplementation(ODataResource resource)\r\n at Microsoft.OData.ODataWriterCore.WriteStart(ODataResource resource)\r\n at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteComplexAndExpandedNavigationProperty(IEdmProperty edmProperty, SelectExpandClause selectExpandClause, ResourceContext resourceContext, ODataWriter writer)\r\n at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteExpandedNavigationProperties(IDictionary`2 navigationPropertiesToExpand, ResourceContext resourceContext, ODataWriter writer)\r\n at System.Web.OData.Formatter.Serialization.ODataResourceSerializer.WriteResource(Object graph, ODataWriter writer, ODataSerializerContext writeContext, IEdmTypeReference expectedType)\r\n at System.Web.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteResourceSet(IEnumerable enumerable, IEdmTypeReference resourceSetType, ODataWriter writer, ODataSerializerContext writeContext)\r\n at Microsoft.Crm.Extensibility.OData.CrmODataResourceSetSerializer.WriteObject(Object graph, Type type, ODataMessageWriter messageWriter, ODataSerializerContext writeContext)\r\n at System.Web.OData.Formatter.ODataMediaTypeFormatter.WriteToStream(Type type, Object value, Stream writeStream, HttpContent content, HttpContentHeaders contentHeaders)\r\n at System.Web.OData.Formatter.ODataMediaTypeFormatter.WriteToStreamAsync(Type type, Object value, Stream writeStream, HttpContent content, TransportContext transportContext, CancellationToken cancellationToken)\r\n--- End of stack trace from previous location where exception was thrown ---\r\n at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n at System.Web.Http.WebHost.HttpControllerHandler.d__1b.MoveNext()"
}
}
}

before ODL 7.0 upgrade it worked fine.

### Assemblies affected
System.Web.OData

### Reproduce steps
PUT on http://kalavany3/test/api/data/v8.2/accounts(c215ae71-9ff8-e611-80d5-00155d224f03)/primarycontactid/$ref

body - 
{
	"@odata.id":"http://kalavany3/test/api/data/v8.2/contacts(3f24cef7-88f8-e611-80d5-00155d224f03)"
	}

### Expected result
It should not throw exception but continue and associate the contact mentioned in the payload to the account.

### Actual result
throwing the exception - 
{
  "error": {
    "code": "",
    "message": "No HTTP resource was found that matches the request URI 'http://kalavany3/test/api/data/v8.2/accounts(c215ae71-9ff8-e611-80d5-00155d224f03)/primarycontactid/$ref'."
  }
}

**Comment by jatin085** [comment]
We should close this issue.

We found the issue was in our handing of navigation path segment after we upgraded to new OData library. Earlier, there was a single type `NavigationPathSegment` in WebApi for both navigation property URI segment and navigation property reference segment. After moving to OData .NET types, navigation property and reference segments are defined by separate types `NavigationPropertySegment` and `NavigationPropertyLinkSegment`. The handling for `NavigationPropertyLinkSegment` was missing on our side. After making the required changes, the issue seems to be fixed.
